### PR TITLE
Add lightweight libcanard based DroneCAN driver for Vehicle builds

### PR DESCRIPTION
**Pull request by bugobliterator**
Following is the list of tests added for this PR:
* Added SITL test (without Virtual CAN) for Compass, Baro, and Airspeed, Test with GPS periph based test with Virtual CAN still supported. 
* UAVCAN sniffer can now be built natively on linux platforms.
* Google test for CubeFramework with and without libcanard.

- [x] Need F4, F7, H7 testing

### CODE REVIEW

File: libraries/AP_DroneCAN/examples/UAVCAN_sniffer/UAVCAN_sniffer.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    START_CB(uavcan_equipment_power_BatteryInfo, BatteryInfo);
+    START_CB(uavcan_equipment_actuator_ArrayCommand, ArrayCommand);
+    START_CB(uavcan_equipment_esc_RawCommand, RawCommand);
+    START_CB(uavcan_equipment_indication_LightsCommand, LightsCommand);
+    START_CB(com_hex_equipment_flow_Measurement, Measurement);
+
+    debug_uavcan("UAVCAN: init done\n\r");
+}
+
+static void send_node_status()
```
Comment (bugobliterator) [inline libraries/AP_DroneCAN/examples/UAVCAN_sniffer/UAVCAN_sniffer.cpp:178]: send node status

File: libraries/AP_DroneCAN/AP_CubeFramework_iface.cpp [hunk 2]
```diff
@@ -148,14 +151,22 @@ void CanardInterface::processTestRx() {
...
+            txf = canardPeekTxQueue(&canard);
+            if (txf == nullptr) {
+                break;
+            }
+            txmsg.dlc = AP_HAL::CANFrame::dataLengthToDlc(txf->data_len);
+            memcpy(txmsg.data, txf->data, txf->data_len);
+            txmsg.id = (txf->id | AP_HAL::CANFrame::FlagEFF);
 #if HAL_CANFD_SUPPORTED
-        txmsg.canfd = txf->canfd;
+            txmsg.canfd = txf->canfd;
```
Comment (bugobliterator) [inline libraries/AP_DroneCAN/AP_CubeFramework_iface.cpp:167]: pull in txf->iface_mask under Semaphore

File: libraries/AP_DroneCAN/AP_CubeFramework_iface.h [hunk 3]
```diff
@@ -62,5 +62,6 @@ class CanardInterface : public CubeFramework::Interface {
     HAL_EventHandle _event_handle;
     uint8_t tx_fail_count;
     bool initialized;
+    HAL_Semaphore _sem;
 };
 #endif // HAL_ENABLE_DRONECAN_DRIVERS
```
Comment (bugobliterator) [inline libraries/AP_DroneCAN/AP_CubeFramework_iface.h:67]: add line

File: libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp [hunk 4]
```diff
@@ -53,46 +45,50 @@ void AP_OpenDroneID::dronecan_init(AP_UAVCAN *uavcan)
...
 
-    dc_operator_id[driver_index] = new uavcan::Publisher<dronecan::remoteid::OperatorID>(*uavcan->get_node());
+    // dc_operator_id[driver_index] = new uavcan::Publisher<dronecan::remoteid::OperatorID>(*uavcan->get_node());
+    dc_operator_id[driver_index] = new CubeFramework::Publisher<dronecan_remoteid_OperatorID_cxx_iface>(uavcan->get_canard_iface());
     if (dc_operator_id[driver_index] == nullptr) {
         goto alloc_failed;
     }
-    dc_operator_id[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
-    dc_operator_id[driver_index]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
+    // dc_operator_id[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
```
Comment (bugobliterator) [inline libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp:83]: add setPriority setTimeout to Publisher

File: libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp [hunk 5]
```diff
@@ -149,12 +145,12 @@ void AP_OpenDroneID::dronecan_send(AP_UAVCAN *uavcan)
 }
 
 #define ODID_COPY(name) msg.name = pkt.name
-#define ODID_COPY_STR(name) do { for (uint8_t i = 0; i<sizeof(pkt.name) && pkt.name[i]; i++) msg.name.push_back(pkt.name[i]); } while(0)
+#define ODID_COPY_STR(name) do { strncpy_noterm((char*)msg.name.data, (const char*)pkt.name, sizeof(msg.name.data)); msg.name.len = strlen((char*)pkt.name); } while(0)
```
Comment (bugobliterator) [inline libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp:148]: strncpy_noterm return copy length

File: libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp [hunk 6]
```diff
@@ -230,13 +226,12 @@ void AP_OpenDroneID::dronecan_send_operator_id(AP_UAVCAN *uavcan)
...
  */
-static void handle_arm_status(AP_UAVCAN* ap_uavcan, uint8_t node_id, const ArmStatusCb &cb)
+static void handle_arm_status(AP_DroneCAN *ap_uavcan, const CanardRxTransfer& transfer, const dronecan_remoteid_ArmStatus &msg)
 {
-    const auto &msg = *cb.msg;
     mavlink_open_drone_id_arm_status_t status {};
 
     status.status = msg.status;
-    strncpy_noterm(status.error, msg.error.c_str(), sizeof(status.error));
+    strncpy_noterm((char*)status.error, (const char*)msg.error.data, sizeof(status.error));
```
Comment (bugobliterator) [inline libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp:234]: strncpy_noterm return copy length

**Comment by davidbuzz** [comment]
I had a look at this, and I'm personally not a fan of moving/hiding a bunch of code over in the submodule.... https://github.com/CubePilot/CubeFramework .  If the submodule stays, it needs to be https://github.com/ArduPilot/xxxxxFramework

Although I appreciate the huge effort gone into writing it, branding it as 'Cube' everywhere isn't the right move for ArduPilot as its fundamentally a canard framework, so calling it 'CANFramework' or 'CanardFramework' or similar is a more accurate description of what it does/ what its for. 

 The submodule/repo, and the code within ,does not specify any GPL or similar licence, so that's also a problem with including it.
 
 I'd rather see the 'framework' bits in libraries/AP_CANManager/ or libraries/AP_CANFramework/ or something like that rather than have the submodule at all.
 
[ There's a *lot* of churn in this code as well, much too hard to review for functional changes in one big hit, but i did what i could, and other than the above gotchas's , i like the general direction its going.   I'm sure there's gonna be more/other feedback too.]

File: .gitmodules [hunk 7]
```diff
@@ -34,3 +34,6 @@
 [submodule "modules/DroneCAN/libcanard"]
 	path = modules/DroneCAN/libcanard
 	url = https://github.com/DroneCAN/libcanard
+[submodule "module/CubeFramework"]
+	path = modules/CubeFramework
+	url = https://github.com/CubePilot/CubeFramework/
```
Comment (tridge) [inline .gitmodules:39]: please move to under DroneCAN repo, and I don't think it should use a vendor specific name like 'Cube'

**Review by tridge — CHANGES_REQUESTED** [review]

File: .gitmodules (hunk 7 above)
Comment (bugobliterator) [reply .gitmodules]: done, I moved the wrappers to libcanard now. That seemed the neatest solution.

File: libraries/AP_Baro/AP_Baro_UAVCAN.cpp [hunk 8]
```diff
@@ -31,23 +26,25 @@ void AP_Baro_UAVCAN::subscribe_msgs(AP_UAVCAN* ap_uavcan)
...
-    uavcan::Subscriber<uavcan::equipment::air_data::StaticPressure, PressureCb> *pressure_listener;
-    pressure_listener = new uavcan::Subscriber<uavcan::equipment::air_data::StaticPressure, PressureCb>(*node);
-    // Msg Handler
-    const int pressure_listener_res = pressure_listener->start(PressureCb(ap_uavcan, &handle_pressure));
-    if (pressure_listener_res < 0) {
-        AP_HAL::panic("UAVCAN Baro subscriber start problem\n\r");
+#if AP_TEST_DRONECAN_DRIVERS
+    get_uavcan_backend(ap_uavcan, 125, true); // add entry for test driver
+#endif
+    auto *pressure_cb = Canard::allocate_arg_callback(ap_uavcan, &handle_pressure);
```
Comment (tridge) [inline libraries/AP_Baro/AP_Baro_UAVCAN.cpp:32]: why is this two calls, first to allocate_arg_callback() then to Canard::Subscriber. That means two lots of allocation failure handling.

File: libraries/AP_Compass/AP_Compass_UAVCAN.cpp [hunk 9]
```diff
@@ -204,4 +221,67 @@ void AP_Compass_UAVCAN::read(void)
     drain_accumulated_samples(_instance);
 }
 
+#if AP_TEST_DRONECAN_DRIVERS
+void AP_Compass_UAVCAN::_setup_eliptical_correcion(uint8_t i)
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_UAVCAN.cpp:225]: I find it strange to have this test code here. Why not in libraries/SITL/ ?

File: libraries/AP_UAVCAN/AP_UAVCAN.h [hunk 10]
```diff
@@ -336,21 +235,68 @@ class AP_UAVCAN : public AP_CANDriver, public AP_ESC_Telem_Backend {
...
+
     // safety status send state
     uint32_t _last_safety_state_ms;
 
     // notify vehicle state
     uint32_t _last_notify_state_ms;
+    uavcan_protocol_NodeStatus node_status_msg;
+
+    CanardInterface canard_iface;
+    Canard::Publisher<uavcan_protocol_NodeStatus_cxx_iface> node_status{canard_iface};
```
Comment (tridge) [inline libraries/AP_UAVCAN/AP_UAVCAN.h:249]: what is the memory cost of these publishers when DroneCAN is not enabled?

**Review by tridge — CHANGES_REQUESTED** [review]
thanks for the restructure!
I had a quick look at the canard C++ wrapper and it doesn't look like any locking is done in the list handling. Is thread safety a todo?

File: libraries/AP_UAVCAN/AP_UAVCAN.h (hunk 10 above)
Comment (bugobliterator) [reply libraries/AP_UAVCAN/AP_UAVCAN.h:249]: There's no cost, we allocate the class only when we use it. Previously, the publishers were static pointers, so technically the memory cost is less than before.

File: libraries/AP_Compass/AP_Compass_UAVCAN.cpp (hunk 9 above)
Comment (bugobliterator) [reply libraries/AP_Compass/AP_Compass_UAVCAN.cpp:225]: Its just replicating, AP_Compass_SITL.cpp

File: libraries/AP_Baro/AP_Baro_UAVCAN.cpp (hunk 8 above)
Comment (bugobliterator) [reply libraries/AP_Baro/AP_Baro_UAVCAN.cpp:32]: I guess I can add another template argument of Callback type to Canard::Subscriber<> if you want.

File: libraries/AP_Baro/AP_Baro_SITL.cpp [hunk 11]
```diff
@@ -29,17 +29,17 @@ AP_Baro_SITL::AP_Baro_SITL(AP_Baro &baro) :
 void AP_Baro_SITL::temperature_adjustment(float &p, float &T)
 {
     const float tsec = AP_HAL::millis() * 0.001f;
-    const float T_sensor = T + _sitl->temp_board_offset;
-    const float tconst = _sitl->temp_tconst;
+    const float T_sensor = T + AP::sitl()->temp_board_offset;
```
[the remark this answers is not in the document]
Comment (bugobliterator) [reply libraries/AP_Baro/AP_Baro_SITL.cpp:32]: I am reusing this method in DroneCAN test.

**Comment by bugobliterator** [comment]
> thanks for the restructure! I had a quick look at the canard C++ wrapper and it doesn't look like any locking is done in the list handling. Is thread safety a todo?

Yes, thread safety is a todo. We subscribe to messages inside single function call and only do it once. And we never delete subscribers. So I don't think thread safety is required for Ardupilot, yet.

File: libraries/AP_Compass/AP_Compass_UAVCAN.cpp (hunk 9 above)
Comment (bugobliterator) [reply libraries/AP_Compass/AP_Compass_UAVCAN.cpp]: I have moved all such tests out of drivers and into libraries/SITL as discussed.

**Comment by tridge** [comment]
I flew this today on a Striver Mini quadplane with CubeOrange. I had DroneCAN airspeed, baro, GPS and compass

**Comment by ukmook** [comment]
I flew this yesterday on a chimera 7" quad, equipped with DroneCan gps, baro, mag, alarm/buzzer and NCP5623 RGB led before changes were made, i have updated and will test again.

**Comment by bugobliterator** [comment]
> I flew this yesterday on a chimera 7" quad, equipped with DroneCan gps, baro, mag, alarm/buzzer and NCP5623 RGB led before changes were made, i have updated and will test again.

Thanks @ukmook !

**Comment by tridge** [comment]
@bugobliterator can you rebase? We have branched 4.4 so this PR can now be merged when it is passing (and works!)

**Review by tridge — APPROVED** [review]

**Comment by peterbarker** [comment]
Ping @bugobliterator @magicrub this PR broke support for logging of Volz ActuatorStatus (no replacement for the loss of com::volz::servo::ActuatorStatus subscriber:
image

**Comment by magicrub** [comment]
Yeah, I fixed it on my 4.5 fork but haven't gotten around to PR it. I'll do it soon.

**Pull request merged**